File: solution/java/week03-cqs/exercise2/src/main/java/cache/Cache.java
package cache;

import java.util.HashMap;
import java.util.Map;

public class Cache {
    private final Map<String, Integer> map = new HashMap<>();

    public Integer get(String key) {
        return map.get(key); // Just returns the value or null if not present
    }

    public void insertIfAbsent(String key, int value) {
        map.putIfAbsent(key, value); // Only modifies the state, does not return value
    }

    public int ensurePresentAndGet(String key, int defaultValue) {
        insertIfAbsent(key, defaultValue);
        return get(key);
    }
}

File: solution/kotlin/week03-exercise2/src/main/kotlin/cache/Cache.kt
package cache

class Cache {
    private val map: MutableMap<String, Int> = mutableMapOf()

    fun get(key: String): Int? = map[key] // Just returns the value or null if not present

    fun insertIfAbsent(key: String, value: Int) {
        map.putIfAbsent(key, value) // Only modifies the state, does not return value
    }

    fun ensurePresentAndGet(key: String, defaultValue: Int): Int {
        insertIfAbsent(key, defaultValue)
        return get(key) ?: defaultValue
    }
}
